### ISSUE TITLE
Add canvas connecting lines for external Simplemind images that don't exist in smmx files

### ISSUE THREAD

**Issue by djsudduth** [djsudduth/mindmap-markdown#9]
When images are added to Simplemind mindmaps - there is an option to either embed or have external images. When external images are used, the smmx file does not have any connectors between the nodes and images. Find a way to add external connection lines on the Obsidian canvas.

**Comment by Hela-Girl** [comment]
@djsudduth  Thanks for adding to the new conversation.

**Comment by djsudduth** [comment]
@Hela-Girl - our biggest problem with this addition is that Obsidian doesn't have straight or dotted connecting lines between cards. So the lines to images will have to be curved and solid.

**Comment by Hela-Girl** [comment]
https://i.imgur.com/TMCojt7.jpeg

For some reason, no connections with the pictures or the pictures themselves appeared. What am I doing wrong?

python mindmd.py -c -t

**Comment by Hela-Girl** [comment]
https://i.imgur.com/ncbTA16.jpeg

Hurray, the cards have appeared, but there is no connection to the pictures.

**Comment by Hela-Girl** [comment]
Maybe you can release it with the same curves?

**Comment by Hela-Girl** [comment]
If it is possible for cards to use “direct connections” it would be great.

**Comment by djsudduth** [comment]
We'll see - not sure just how it will work yet. The current 0.2.0 doesn't have the connectors yet.

**Comment by djsudduth** [comment]
Version 0.2.1 is ready with the lines / connections to images now added! Again, use the `-c` and `-t` switches:

`python mindmd.py -c -t`

Download using the same link from the 0.2.0 version Here -> https://github.com/djsudduth/mindmap-markdown/tree/develop

Choose the Code button / download to get the .zip file of the new beta version

**Comment by Hela-Girl** [comment]
Minor edits:

1. when converting, is it possible to use jpg or webp instead of png?
2. I don’t know or it’s a mistake, but if the images were originally not in English, then after conversion they are not there.

https://i.imgur.com/lcwvgSB.jpeg

**Comment by djsudduth** [comment]
for your edits:

1. It might be possible to change the image format - PNG is the format coming out of Simplemind - it will require a new issue to investigate using the Pillow utility - see #12 
2. I'll take a look - it's probably a unicode language issue.

Thanks for testing it!

**Comment by djsudduth** [comment]
PS - could you send me the Simplemind smmx file in the image above?

**Comment by djsudduth** [comment]
Ok, I fixed number 2 in your list. It was just a naming bug.  Just download the latest 0.2.1 again from the link above.

**Comment by Hela-Girl** [comment]
https://www.mediafire.com/file/rlmtip4amav7qtx/111.zip/file
an example in which the names of the attached images are not in English.

**Comment by djsudduth** [comment]
This worked fine! All is good.  
Download using the same link from the 0.2.1 version Here -> https://github.com/djsudduth/mindmap-markdown/tree/develop

Choose the Code button / download to get the .zip file of the new beta version

![Map](https://github.com/user-attachments/assets/db2cded1-a902-4a5c-883e-07f10f4442c7)

**Comment by Hela-Girl** [comment]
1. I noticed a minor error from the "images" folder: it does not transfer images to the "media" folder.
2. There are images that are inserted into the block for them, can they be taken if they cannot be inserted into the block the same way?

https://i.imgur.com/H8G5pae.jpeg

https://i.imgur.com/883wvxg.jpeg

**Comment by djsudduth** [comment]
Ok!! So glad you found 1. - missing images in the media folder. It happened to me once and I thought I was imagining it. But, there wasn't an error message - yet the media folder was empty.  I'll look into that.

I'm not quite sure what you mean by 2. Obsidian cards cannot have images within them unless they are markdown notes.  So, if images are within a Simplemind node, they are placed just next to the card unless you choose to have all Simplemind nodes be markdown files.  Can you explain more?

**Comment by Hela-Girl** [comment]
2. You are very insightful. You understood everything correctly.
---
Please tell me if you could put the necessary module for converting images into jpg / png format in the folder with your plugin. So that when converting, people don’t have to search for and download the required module? (It’s not clear what to just do the conversion without additional searches 😄)

**Comment by Hela-Girl** [comment]
P.S.: It’s a pity that obsidian has so few possibilities for connections/lines and possible manipulations with lines. SimpleMind has a much better release in this regard.

**Comment by djsudduth** [comment]
I tried to reproduce 1 - images not moving to media. They all transfer as far as I can tell. But, if there are a lot of images you have to restart Obsidian to see them all - you can't leave Obsidian running or the images won't all refresh on the screen making it look like images are missing.

**Comment by Hela-Girl** [comment]
I don’t know or I understood you correctly, but the images from the “images” folder can simply be copied with a script to the “media” folder, as I did it manually. And it worked for me.

**Comment by djsudduth** [comment]
I was just saying that whenever I run the program, all of the files are moved to the "media" folder successfully. I've tried it with your samples and all worked just fine.  However, in order to see them in Obsidian, you have to shut Obsidian down and restart it.

**Comment by Hela-Girl** [comment]
Please tell me where to download the updated version of the script with the removal of “media” and images in the topic (within the topic). Or is it still being developed?

**Comment by Hela-Girl** [comment]
And I also noticed an error occurs if the file name *.smmx is not in English. The example "Позы.smmx" gives an error, but if you rename it to Pose.smmx, everything works fine.

**Comment by djsudduth** [comment]
@Hela-Girl - I don't understand what you mean by "removal of “media” and images in the topic (within the topic)"
What are you looking for?  Right now, if you embed media within the topic, the image will appear either inside the markdown note if not using the cards. If you are using the card view then the image is placed outside the card next to it.

**Comment by Hela-Girl** [comment]
https://i.imgur.com/kugIsDV.jpeg

1.  In the SimpleMind program, an image can be inserted into a “topic” (1 image), but Obsidian cannot do this. Was there a question or did you correct the fact that images in Obsidian will be linked to the topic?

2. After the script has finished running, there is a “media” folder in the “betamarkdown” folder and it is always empty for me. I move the images from the "images" folder to the "media" folder and everything works fine for me. (rebooting Obsidian) Question or did you fix this issue?

**Comment by Hela-Girl** [comment]
P.S.: Sorry, I’m a little confused with the edits that have been made or those that will be made.

**Comment by djsudduth** [comment]
For 1: If you export a simplemind file that has an embedded image to an Obsidian canvas with only cards, the embedded image is extracted and placed next to the card in Obsidian. That is working now.

For 2: Check your Settings.cfg file to make sure the images are going to the `media` folder (I think there was a mistake in one of the beta versions that changed the config and put the images in `/ddd/media` folder accidently).  Here's what the default Settings should be:

```
[SETTINGS]
input_path = mindmaps/
output_path = markdown/
media_path = media/
test_file_name = HII Regions.smmx
canvas_scale = 2.2
obsidian_vault_name = Obsidian
```

Here's how I have mine setup to publish directly to the Obsidian vault:
```
[SETTINGS]
input_path = mindmaps/
output_path = C:/Users/Username/Google Drive/Obsidian/Canvas
media_path = media/
test_file_name = HII Regions.smmx
canvas_scale = 2.2
obsidian_vault_name = Obsidian
```

**Comment by Hela-Girl** [comment]
[SETTINGS]
input_path = mindmaps/
output_path = betamarkdown/
media_path = media/
test_file_name = 3333.smmx
canvas_scale = 1
obsidian_vault_name = Obsidian/

These are my settings:

**Comment by djsudduth** [comment]
So, any time you run the program, the images are not in the `betamarkdown/media` folder?

**Comment by Hela-Girl** [comment]
no only in "images"

**Comment by djsudduth** [comment]
Can you look within your `betamarkdown/` folder and see if the images are somewhere else within it?

What OS / Version are you running?  I'm running on Windows 10.

**Comment by Hela-Girl** [comment]
Windows 10
There are definitely no images in the betamarkdown/ folder after conversion.

**Comment by djsudduth** [comment]
Is there a `media` folder within the `betamarkdown` folder?  If not, try creating it and run the program.

**Comment by Hela-Girl** [comment]
the media folder is always in the betamarkdown folder

**Comment by djsudduth** [comment]
Ok - I changed the copying of files to report any errors from the images to the media folders. I also fixed the error of .smmx not being in English.

Download the new beta version Here -> https://github.com/djsudduth/mindmap-markdown/tree/develop

Choose the Code button / download to get the .zip file of the new beta version.

**Comment by Hela-Girl** [comment]
https://i.imgur.com/FKrgJIP.jpeg
Something is wrong....

An idea came to me: In the settings "test_file_name = Poses.smmx" is it possible to do without specifying the file for conversion? You can, after all, any file that is in the "mindmaps" folder can be converted. Just before starting, it will show which file was found in the folder and ask a question or whether to convert this file yes/no.

**Comment by djsudduth** [comment]
Hi! For your 2 questions / issues:

1. I think the problem with the file copying has something to do with the .smmx file getting unzipped (extracting the files out). I think the program is running faster than the extraction so it thinks the files are missing.  Is your Windows 10 computer a newer very fast one with SSD drive? Or, is your computer an older slower one pre-2018?
2. Yes, your idea is possible. If you use the `-d` flag in outline mode (not canvas or card view), all of the .smmx files are converted. I haven't added batch yet for canvas output but will consider your idea.

**Comment by Hela-Girl** [comment]
Yes, the computer is indeed very fast. Two NVMe (M.2) ssd work in a raid to increase speed.

However, this can be solved by waiting for the conversion program to finish its work completely and then simply transferring the image files from one folder to another.

**Comment by djsudduth** [comment]
Ok - that answers it. The file extraction is slower than the program - so the files aren't fully extracted by the time the program is done (that's why the files are missing - unzipping the files is multi-threaded). I'm going to changed the unzipping of the smmx files to wait until they are all done. Also, slow do the copying to make sure the file is copied to media.

**Comment by Hela-Girl** [comment]
1. Could you change it in the script so that it is copied after the script is completed?
2. Can I ask you to add it so that it works for canvas?
( > If you use the -d flag in outline mode (not canvas or card view), all of the .smmx files are converted. I haven't added batch yet for canvas output but will consider your idea.)

**Comment by djsudduth** [comment]
Ok - I made a change to extract the smmx zip file a single files at a time rather than unzipping / extracting the entire smmx file at once. It will also check to see if the files are written to disk.
(If this doesn't work I may try what you suggested in 1)

I'm not ready to add batch for canvases yet - but will add a new issue for that

Download the new fix version Here -> https://github.com/djsudduth/mindmap-markdown/tree/develop

Choose the Code button / download to get the .zip file of the new fix version.

**Comment by Hela-Girl** [comment]
https://i.imgur.com/YSX0Sql.jpeg

Something is wrong....

**Comment by djsudduth** [comment]
Download it again and try it again now

**Comment by Hela-Girl** [comment]
https://i.imgur.com/fGi7XTD.jpeg

The same...

**Comment by djsudduth** [comment]
Interesting - only one image file had the error this time??

**Comment by djsudduth** [comment]
Another change - I've added more retries and delays between copying files.

Download it again and try it again now

**Comment by Hela-Girl** [comment]
https://i.imgur.com/mc6Pbd4.jpeg

**Comment by djsudduth** [comment]
Can you please try some different smmx files and different paths and see if that problem continues?

**Comment by djsudduth** [comment]
Also, please be sure the images are not open in any other programs like image viewers or in Obsidian

**Comment by Hela-Girl** [comment]
No, not open anywhere. I think the problem is in the algorithm needed after the script is completed. Only then copy the images to another folder and not one by one.

**Comment by djsudduth** [comment]
Can you verify that the file - Image file `images/7a2baf2b08ef6d8e06cd618a2986dc7cba914d2e.png`  actually exists in the `images` folder when you run the program (and be sure to erase all files in the `images` folder before running the program)?

Did you try other smmx files?  Also, I never did ask you what version of Python you're using. I'm running 3.12.  Be sure you're on at least 3.11.

The main problem is that file handling both for zip and copying files is done multi-threaded in Python. Your system is so fast and the raid configuration is most likely doing write caching that is blocking the file copy.

**Comment by Hela-Girl** [comment]
Python 3.13.0
1. In the script you can forcefully empty the "images" folder. Like the "betamarkdown" folder.
2. I processed this smmx file with your script before and everything worked correctly.

**Comment by djsudduth** [comment]
Did that smmx file work with the current 0.1.0 release version?  

Also, please try a fresh new beta directory with some other smmx files. 

All of the sample files you've sent me work with no errors here in both the newest 0.2.2 and 0.1.0 versions.

**Comment by Hela-Girl** [comment]
Everything worked until the moment when files began to be transferred automatically.

**Comment by djsudduth** [comment]
I don't know what you mean. They've always been transferred / copied to the media folder since it was first developed.

**Comment by Hela-Girl** [comment]
https://github.com/djsudduth/mindmap-markdown/issues/9#issuecomment-2546134752

**Comment by djsudduth** [comment]
I've added the original version 0.0.0 to the files called `oldmindmd.py`.  This is the first version that didn't have any of the Canvas work. All the versions since I first developed it always copied the images from the unzip `images\` folder to the default `media\` folder. You can try that to see if there are any differences - it will be interesting to see what happens. 

Download it here -> https://github.com/djsudduth/mindmap-markdown/tree/develop

Choose the Code button / download to get the .zip file of the new version with that file.  You run it with `python oldmindmd.py`

**Comment by Hela-Girl** [comment]
Can you restore the version from two days ago?

**Comment by djsudduth** [comment]
For your comment https://github.com/djsudduth/mindmap-markdown/issues/9#issuecomment-2546134752 above, the images files should **always** be in the media folder. If they are not, then something is most likely wrong in `settings`.

Please use the default `settings.cfg` - not the one you modified.  Just install the new 0.2.2 version in a new, fresh directory and do not change the settings at all. Run the sample test file with just `python mindmd.py` and confirm if the images are in the `markdown/media` folder within the install directory. There should be 5 image files in the media folder from the sample test HII Regions.smmx file!

Please let me know what you discover.

**Comment by Hela-Girl** [comment]
Hurray, I solved the problem, the “media” folder was missing, I added it and all the missing whites now appeared))
Now it transfers the images to the "media" folder but does not delete them from the "images" folder.

**Comment by Hela-Girl** [comment]
I wanted to express my deep gratitude for helping me with the transfer of projects from SimpleMind to Obsidian for almost a year! I understand that it’s still early, but still happy holidays and thank you again for your work!

**Comment by djsudduth** [comment]
Glad that's working out for you! Appreciate your message!

**Comment by djsudduth** [comment]
Added in version 0.2.2 dev in 12/2024 - now in release 0.2.3
